tests/cache/test_mesi_controller.py

```diff
@@ -1,116 +1,479 @@
+        """
+        Test handling a snooped bus read operation for a cacheline in INVALID state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.INVALID, BusOp.READ, address)
+
+        # Check that the snoop result was put
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.NOHIT)
+
+        # Assert state transition: INVALID -> INVALID
+        self.assertEqual(next_state, MESIState.INVALID)
+
+    def test_snoop_invalid_bus_write(self):
+        """
+        Test handling a snooped bus write operation for a cacheline in INVALID state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.INVALID, BusOp.WRITE, address)
+
+        # Check that the snoop result was not put
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.NOHIT)
+
+        # Assert state transition: INVALID -> INVALID
+        self.assertEqual(next_state, MESIState.INVALID)
+
+    def test_snoop_invalid_bus_rwim(self):
+        """
+        Test handling a snooped bus RWIM operation for a cacheline in INVALID state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.INVALID, BusOp.RWIM, address)
+
+        # Check that the snoop result was not put
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.NOHIT)
+
+        # Assert state transition: INVALID -> INVALID
+        self.assertEqual(next_state, MESIState.INVALID)
+
+    def test_snoop_invalid_bus_invalidate(self):
+        """
+        Test handling a snooped bus invalidate operation for a cacheline in INVALID state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.INVALID, BusOp.INVALIDATE, address)
+
+        # Check that the snoop result was not put
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.NOHIT)
+
+        # Assert state transition: INVALID -> INVALID
+        self.assertEqual(next_state, MESIState.INVALID)
+
+    def test_snoop_shared_bus_read(self):
+        """
+        Test handling a snooped bus read operation for a cacheline in SHARED state
+        """
+        address = 0x1000
+
+        # Call the function under test
+        next_state = self.controller.handle_snoop(MESIState.SHARED, BusOp.READ, address)
+
+        # Check that the snoop result was not put
+        self.mock_put_snoop.assert_called_with(address, SnoopResult.HIT)
+
+        # Assert state transition: SHARED -> SHARED
+        self.assertEqual(next_state, MESIState.SHARED)
+
+    def test_snoop_shared_bus_write(self):
+        """
+        Test handling a snooped bus write operation for a cacheline in SHARED state, 
+        this would never happpen under MESI protocol
```

Comment:
Just noting, I had another comment that relates to this one. Namely that I think we might get by with an explanatory comment regarding write bus ops, rather than including any tests.

Reply:
It's important to test cases with an invalid input and confirm that the state transition doesn't do anything unexpected. This is an example of a scenario that wouldn't happen under MESI but could happen in a trace file.

Reply:
Yeah, as I mentioned above, I think if we're working under the assumption we're supposed to handle an invalid ordering of events from a trace file, we'd probably want to add an error or warning from the MESI controller, or wherever we are in the code where we can detect such an invalid state. Maybe we can clarify with Mark how he wants us to handle it.

Reply:
@nkanderson I added runtime warnings for these "not possible" events as we discussed last week.